client/src/components/CommentApp/comments.js

```diff
@@ -0,0 +1,144 @@
+    // and that screenreader users can return to the annotation point when desired
+  }
+  show() {
+    this.node.classList.remove('u-hidden');
+  }
+  hide() {
+    this.node.classList.add('u-hidden');
+  }
+  setOnClickHandler(handler) {
+    this.node.addEventListener('click', handler);
+  }
+  getDesiredPosition() {
+    return (
+      this.fieldNode.getBoundingClientRect().top +
+      document.documentElement.scrollTop
+    );
+  }
+}
+
+class FieldLevelCommentWidget {
+  constructor(
+    fieldNode,
+    commentAdditionNode,
+    annotationTemplateNode,
+    commentApp
+  ) {
+    const self = this;
+    this.fieldNode = fieldNode;
+    this.contentPath = getContentPath(fieldNode);
+    this.commentAdditionNode = commentAdditionNode;
+    this.commentApp = commentApp;
+    this.annotationTemplateNode = annotationTemplateNode;
+    commentAdditionNode.addEventListener('click', self.addComment.bind(self));
+    this.commentNumber = 0;
+    this.commentsEnabled = false;
+  }
+  register() {
+    // if the widget has a contentpath - ie commenting is enabled for this field -
+    // register the widget with the comment app to subscribe to updates on comments for
+    // this widget's contentpath, and to commenting enabled/disabled updates
+    if (this.contentPath) {
+      this.commentApp.registerWidget(this);
+    }
+  }
+  setEnabled(enabled) {
+    // Update whether comments are enabled for the page
+    this.commentsEnabled = enabled;
+    this.updateVisibility();
+  }
+  onChangeComments(comments) {
+    // Receives a list of comments for the widget's contentpath
+    this.commentNumber = comments.length;
+    this.updateVisibility();
+  }
+  updateVisibility() {
+    // if comments are disabled, or the widget already has at least one associated comment,
+    // don't show the comment addition button
+    if (!this.commentsEnabled || this.commentNumber > 0) {
+      this.commentAdditionNode.classList.add('u-hidden');
+    } else {
+      this.commentAdditionNode.classList.remove('u-hidden');
+    }
+  }
+  getAnnotationForComment() {
+    const annotationNode = this.annotationTemplateNode.cloneNode(true);
+    annotationNode.id = '';
+    this.commentAdditionNode.insertAdjacentElement('afterend', annotationNode);
+    return new BasicFieldLevelAnnotation(this.fieldNode, annotationNode);
+  }
+  addComment() {
+    this.commentApp.makeComment(
+      this.getAnnotationForComment(),
+      this.contentPath
+    );
+  }
+}
+
+function initFieldLevelCommentWidget(fieldElement) {
+  const widget = new FieldLevelCommentWidget(
+    fieldElement,
+    fieldElement.querySelector('[data-comment-add]'),
+    document.querySelector('#comment-icon'),
+    window.commentApp
+  );
+  widget.register();
+}
+
+function prepareFieldLevelCommentInit(fieldSelector) {
+  if (window.commentApp) {
+    initFieldLevelCommentWidget(fieldSelector);
+  } else {
+    document.addEventListener('DOMContentLoaded', () => {
```

Comment:
JS in general can run before it, but you're right, since it's passed an element anyway, doing it after `DOMContentLoaded` is probably implicit! I'll try removing it.

Reply:
Removing caused errors due to widgets initialising themselves before. I agree the two step init isn't great though, so I've tried replacing with callbacks in the latest commits.